teps/0141-platform-context-variables.md

```diff
@@ -0,0 +1,425 @@
+  - name: build
+    taskRef:
+      resolver: hub
+      params:
+      - name: name
+        value: kaniko
+      - name: version
+        value: "0.6"
+    params:
+    - name: image
+      value: $(params.image)
+    workspaces:
+    - name: source-code
+    runAfter:
+    - fetch-source
+```
+
+The platform builder may want to provide default variable substitutions for these parameters, or parts of these parameters. For example, the platform could provide a variable substitution for the URL of a repo where an event triggering a run of this Pipeline occurred, which could be used to substitute the full "repo_url" param. The platform could also provide the location of the image/artifact registry where the built image should be pushed, which could be used to substitute part of the "image" param.
+
+This section discusses several ways the platform builder could provide these parameter substitutions.
+
+#### Substitute directly in PipelineRun spec
+
+The platform could directly substitute any variables defined in the PipelineRun spec before submitting the PipelineRun to their Tekton implementation. For example, let's say the user defines the following PipelineRun:
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: PipelineRun
+metadata:
+  name: clone-kaniko-build-push-run
+spec:
+  pipelineRef:
+    name: clone-kaniko-build-push
+  params:
+  - name: repo-url
+    value: $REPO_URL
+  - name: image
+    value: $IMAGE_REGISTRY/myapp
+  workspaces:
+  - name: source-code
+    ...
+```
+
+The platform builder could substitute `$REPO_URL` and `$IMAGE_REGISTRY` with the appropriate values before creating the PipelineRun.
+However, this would involve mutating user-provided PipelineRun specs, and would likely not be considered Tekton conformant, as PipelineRuns don't support the syntax `$REPO_URL` and `$IMAGE_REGISTRY`. It would also cause confusion if PipelineRuns did start supporting this syntax in the future.
+
+#### Build a higher-level API
+
+To avoid mutating user-specified PipelineRun specs, platforms could build higher-level APIs that reference PipelineRuns, and allow variable substitutions within these APIs. This is the approach taken by Triggers. For example, TriggerTemplates support the following syntax:
+
+```yaml
+apiVersion: triggers.tekton.dev/v1beta1
+kind: TriggerTemplate
+metadata:
+  name: clone-build-template
+spec:
+  params:
+  - name: repo-url
+  - name: image-registry
+  resourcetemplates:
+  - apiVersion: tekton.dev/v1
+    kind: PipelineRun
+    metadata:
+      generateName: clone-build-run-
+    spec:
+      pipelineRef:
+        name: clone-kaniko-build-push
+      params:
+      - name: repo-url
+        value: $(tt.params.repo-url)
+      - name: image-registry
+        value: $(tt.params.image-registry)/myapp
+```
+
+This approach adds verbosity for PipelineRun authors, and doesn't allow provenance generators to distinguish between parameters provided by the PipelineRun author from parameters provided by the platform.
```

Comment:
One question we should ask oursleves is : who are `PipelineRun` authors ? 
Because, I think in a majority of cases, "they" would be "robots" 😛

Reply:
I think its useful to think in terms of user provided inputs (eg. coming from event/user params/another system or robot etc.) vs system provided inputs that are trusted and provided by Tekton itself

Reply:
I agree with the general idea that we have two types of users that need different capabilities, but I think part of the problem here is "system provided" can be a grey area - e.g. a triggering system that's providing automation will look like another user w.r.t. the pipelines API, even though to external users it might be considered a trusted part of the overall CI system.

Was chatting w/ @lbernick about this this morning - I think it would make sense to try and model this in RBAC and have a separate verb for these kinds of "elevated" permissions - this would let us be agnostic to what is considered "system provided", and rely on permissions and the capabilities they provide. e.g. users with `create` permissions may be able to create Runs as they can today, but you can only include a `context` if you also have the `create-context` permission (or w/e we want to call it).

Related to this, I think it would be very useful to capture the user who create a Run, since we don't have a good way to determine if a Run created by an automation/system user or not without involving the k8s Audit API. Created https://github.com/tektoncd/pipeline/issues/7068 to track this separately.